### PR TITLE
Studio: Replace button loading indicator with full window one

### PR DESCRIPTION
**Pull request by katinthehatsite**
Related to https://github.com/Automattic/dotcom-forge/issues/8167

## Proposed Changes

This PR replaces the loading state on the `Add site` modal with a full screen loading indicator:

Screenshot 2024-07-16 at 9 02 33 AM

## Testing Instructions

**From "Add site" modal:**

* Pull the changes from this branch
* Start the app with `nvm use && npm install && npm start`
* Click on the `Add site` button to open the modal
* Click on `Add site` to add a site
* Confirm that you see the loading state that takes the whole screen
* Confirm that you see the loading indicator in the sidebar
* Confirm that when you switch to a different site, you see the content tabs as expected
* Confirm that the `Add site` button is blocked for a given site when the site is being created

**From "Onboarding" screen**:

* Pull the changes from this branch
* Start the app with `nvm use && npm install && npm start
* Delete all the sites that you have to trigger the onboarding screen
* Click on `Add site`
* Confirm that you see the loading state that takes the whole screen
* Confirm that you see the loading indicator in the sidebar

**Testing for an error**:

* Apply the following diff:

```
diff --git a/src/ipc-handlers.ts b/src/ipc-handlers.ts
index 95903bb..b2c85d3 100644
--- a/src/ipc-handlers.ts
+++ b/src/ipc-handlers.ts
@@ -130,6 +130,9 @@ export async function createSite(
 ): Promise< SiteDetails[] > {
        const userData = await loadUserData();
        const forceSetupSqlite = false;
+
+       // Introduce an error for testing
+       throw new Error( 'Test Error: This is a forced error for testing purposes.' );
        // We only recursively create the directory if the user has not selected a
        // path from the dialog (and thus they use the "default" or suggested path).
        if ( ! ( await pathExists( path ) ) && path.startsWith( DEFAULT_SITE_PATH ) ) {
```

* Start the app with `nvm use && npm install && npm start`
* Click on the `Add site` button to open the modal
* Click on `Add site` to add a site
* Observe that you see the loading state
* Observe that it throws the dialog with an error
* Observe that the loading state goes away and the site is not created

**Notes**:
* I am not sure what the error handling is supposed to be like according to the design at the moment. For now, I implemented something that seems to make sense and does not look too strange for the user experience. We can open another PR once we have some more clarify on what the design and flow for the error + loading state should look like
* I am planning to more test coverage in a different PR to make this code available for the PR about import interface as soon as possible

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/components/site-loading-indicator.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		setProgress( 80 );
+	}, [ setProgress ] );
+
+	return (
+		<div className="flex flex-col w-full h-full app-no-drag-region pt-8 overflow-y-auto justify-center items-center">
+			<div className="w-[300px] text-center">
+				<div className="text-black a8c-subtitle-small mb-4">{ selectedSiteName }</div>
+				<ProgressBar value={ progress } maxValue={ 100 } />
+				<div className="text-a8c-gray-70 a8c-body mt-4">{ __( 'Creating site...' ) }</div>
+			</div>
```
Comment (katinthehatsite) [inline src/components/site-loading-indicator.tsx:35]: I added `Creating site...` here for now as this is for adding a site. I am expecting that the text would be different when the importing functionality is implemented

File: src/components/add-site.tsx [hunk 2]
```diff
@@ -117,16 +118,20 @@ export default function AddSite( { className }: AddSiteProps ) {
...
+							>
 								{ __( 'Cancel' ) }
 							</Button>
 							<Button
 								type="submit"
 								variant="primary"
-								isBusy={ isAddingSite }
-								disabled={ isAddingSite || !! error || ! siteName?.trim() }
+								isBusy={ selectedSite?.isAddingSite }
+								disabled={ selectedSite?.isAddingSite || !! error || ! siteName?.trim() }
```
Comment (sejas) [inline src/components/add-site.tsx]: I wonder if we can remove the busy and disabled states when `selectedSite?.isAddingSite`.
Or if we want to keep the same logic and limit to create only one site at a time, we should check across all the sites if any of them has the `isAddingSite`

File: src/components/add-site.tsx [hunk 3]
```diff
@@ -117,16 +118,20 @@ export default function AddSite( { className }: AddSiteProps ) {
...
 							>
-								{ isAddingSite ? __( 'Adding site…' ) : __( 'Add site' ) }
+								{ __( 'Add site' ) }
```
Comment (sejas) [inline src/components/add-site.tsx:132]: I've noticed that in Onboarding line 125, we are checking for selectedSite?.isAddingSite to display a different string.
I would apply the same decision as above.

```
{ selectedSite?.isAddingSite ? __( 'Adding site…' ) : __( 'Add site' ) }
```

File: src/hooks/use-add-site.ts [hunk 4]
```diff
@@ -51,22 +50,13 @@ export function useAddSite() {
...
 			await createSite( path, siteName ?? '' );
 		} catch ( e ) {
 			Sentry.captureException( e );
-			setError(
-				__(
-					'An error occurred while creating the site. Verify your selected local path is an empty directory or an existing WordPress folder and try again. If this problem persists, please contact support.'
-				)
-			);
-			setIsAddingSite( false );
-			throw e;
```
Comment (sejas) [inline src/hooks/use-add-site.ts:66]: I see this error has been moved to `useSiteDetails` `createSite` 👍

File: src/hooks/use-site-details.tsx [hunk 5]
```diff
@@ -159,11 +160,50 @@ export function SiteDetailsProvider( { children }: SiteDetailsProviderProps ) {
...
+				if ( newSite?.id ) {
+					setSelectedSiteId( newSite.id ); // Update the selected site to the new site's ID
+				}
+				setData( ( prevData ) =>
+					sortSites(
+						prevData.map( ( site ) =>
+							site.id === tempSiteId ? { ...site, ...newSite, isAddingSite: false } : site
+						)
+					)
+				);
```
Comment (sejas) [inline src/hooks/use-site-details.tsx]: Here I suggest one optimization and one fix.
The optimization is to avoid sortingSites again, since it just check the names. And avoid overwriting the ...site, ...newSite and just use the new site.

The bug fix is that if I create a new site, and I select another site in the sidebar, I was switched to the new site automatically.

```suggestion
				if ( ! newSite ) {
					throw new Error( 'Failed to create site' );
				}
				// Update the selected site to the new site's ID if the user didn't change it
				setSelectedSiteId( ( prevSelectedSiteId ) => {
					if ( prevSelectedSiteId === tempSiteId ) {
						return newSite.id;
					}
					return prevSelectedSiteId;
				} );
				setData( ( prevData ) =>
					prevData.map( ( site ) => ( site.id === tempSiteId ? newSite : site ) )
				);
```

If we apply the suggestion, we need to update the `useSelectedSite` hook too:
```
function useSelectedSite( firstSiteId: string | null ) {
	const SELECTED_SITE_ID_KEY = 'selectedSiteId';
	const selectedSiteIdFromLocal = localStorage.getItem( SELECTED_SITE_ID_KEY ) || null;
	const [ selectedSiteId, setSelectedSiteId ] = useState< string | null >(
		selectedSiteIdFromLocal
	);
	useEffect( () => {
		if ( selectedSiteId ) {
			localStorage.setItem( SELECTED_SITE_ID_KEY, selectedSiteId );
		}
	} );

	return {
		selectedSiteId: selectedSiteId || firstSiteId,
		setSelectedSiteId,
	};
}
```

**Review by sejas — COMMENTED** [review]
I tested it and the happy path works great:

https://github.com/user-attachments/assets/664e1208-835b-4b79-8263-a324cec0a4eb

I added a couple of suggestions and a small fix for an edge case where we selected the new site even if the user would have changed it.

Here is the video for the edge case:

https://github.com/user-attachments/assets/ea33a2ff-f6dd-49c7-8ce3-e8c4f9361674

File: src/hooks/use-add-site.ts (hunk 4 above)
Comment (katinthehatsite) [reply src/hooks/use-add-site.ts:66]: Yeah I moved it over as I did not want it to be a part of the add site form anymore and wanted to trigger it when the loading state happens.

File: src/components/add-site.tsx (hunk 3 above)
Comment (katinthehatsite) [reply src/components/add-site.tsx:132]: Updated in https://github.com/Automattic/studio/pull/362/commits/31548b962f8132d461be5ebf08e3407f321d4203 - thanks for catching this!

File: src/components/add-site.tsx (hunk 2 above)
Comment (katinthehatsite) [reply src/components/add-site.tsx]: Yeah I am not sure if it should allow adding multiple sites at once or not. I think perhaps we can keep the existing logic for now and change in the future if needed. I will add a check for disabling the button when one of the sites is adding a new site, I think

File: src/hooks/use-site-details.tsx (hunk 5 above)
Comment (katinthehatsite) [reply src/hooks/use-site-details.tsx]: Sounds good, I made some changes in https://github.com/Automattic/studio/pull/362/commits/566ef6f2cd5fe2133a8eb97a81701894deff470f and https://github.com/Automattic/studio/pull/362/commits/4bbe32da1993453af41ce5ab18cce25c1ca3e30e to adjust this based on the feedback.

File: src/components/add-site.tsx (hunk 2 above)
Comment (katinthehatsite) [reply src/components/add-site.tsx]: I have done the following:

* removed the `busy` and `disabled` from the `Add site` button on onboarding as it won't be needed https://github.com/Automattic/studio/pull/362/commits/8fe0c6c23f6b994d0d7fb103b2df942a0794da23
* added a check to see if there is `adding site` in progress already to disable the button in the modal to keep the current behaviour: https://github.com/Automattic/studio/pull/362/commits/d180b371a7472dfed515a6eac261b4449498275c

File: src/components/add-site.tsx [hunk 6]
```diff
@@ -117,16 +120,16 @@ export default function AddSite( { className }: AddSiteProps ) {
...
+							<Button onClick={ closeModal } disabled={ isSiteAdding } variant="tertiary">
 								{ __( 'Cancel' ) }
 							</Button>
 							<Button
 								type="submit"
 								variant="primary"
-								isBusy={ isAddingSite }
-								disabled={ isAddingSite || !! error || ! siteName?.trim() }
+								isBusy={ isSiteAdding }
+								disabled={ isSiteAdding || !! error || ! siteName?.trim() }
```
Comment (wojtekn) [inline src/components/add-site.tsx:130]: As we immediately close the dialog, do we need `isBusy` and `disabled` props?

File: src/components/site-loading-indicator.tsx (hunk 1 above)
Comment (wojtekn) [reply src/components/site-loading-indicator.tsx:35]: It's good for now. When we work on import, we can refactor to pass the state to display different messages.

**Review by wojtekn — APPROVED** [review]
The code change looks clear and works as expected. It works on Windows as well.

One note regarding UX - as creating a site on Windows takes more time than on MacOS, should we make the progress bar more dynamic to show some extrapolated progress?

File: src/components/add-site.tsx (hunk 6 above)
Comment (katinthehatsite) [reply src/components/add-site.tsx:130]: Yeah, it is for the case when the user might go and open the `Add site` modal from the sidebar while the new site is being added and a loading state is displayed so that they can't add another new site until this process is finished.
The form on the onboarding does not need it because we move away from that screen right into a loading screen.

Alternatively, if we want to allow creating new sites in parallel, we can remove the props.

For this PR, I left the existing flow and logic that is present in the app now (only one new site can be created at a time) but happy to adjust it.

Comment (wojtekn) [reply src/components/add-site.tsx:130]: Thanks, it sounds reasonable.

**Comment by katinthehatsite** [comment]
@wojtekn would you be able to test on Windows again (I made some changes in https://github.com/Automattic/studio/pull/362/commits/d83fd0ddd382b7fe5f2d6af5f61ffc5036d62d9e ). It looks better on Mac and I would like to confirm that there is an improvement on Windows. Thank you!

File: src/components/add-site.tsx [hunk 7]
```diff
@@ -37,6 +38,8 @@ export default function AddSite( { className }: AddSiteProps ) {
 		loadingSites,
 	} = useAddSite();
 
+	const isSiteAdding = data.some( ( site ) => site.isAddingSite );
```
Comment (sejas) [inline src/components/add-site.tsx:41]: Nice!

**Review by sejas — APPROVED** [review]
Thanks for the changes, I tested it again and it looks great.

https://github.com/user-attachments/assets/99f4ab49-9741-4cd3-924e-fd9a3e654568

I also tested the edge case where you create a site and then select another one. 👍

**Comment by wojtekn** [comment]
> @wojtekn would you be able to test on Windows again (I made some changes in https://github.com/Automattic/studio/commit/d83fd0ddd382b7fe5f2d6af5f61ffc5036d62d9e ). It looks better on Mac and I would like to confirm that there is an improvement on Windows. Thank you!

It's better now, as the progress bar progresses for most of the process. We could improve it even more by adjusting the starting position to be closer to 20%, as now it almost immediately goes to ~50%, and we could make the progress slightly slower. Or maybe it should be enough to adjust starting position.

**Comment by katinthehatsite** [comment]
Sounds good @wojtekn - I will add a separate PR to beautify this 👍 Added an issue here: https://github.com/Automattic/dotcom-forge/issues/8340
